bulbs/poll/models.py

```diff
@@ -0,0 +1,220 @@
+    default_detail = "Third-party poll provider temporarily unavailable."
+
+    def __init__(self, detail):
+        self.detail = detail
+
+class SodaheadResponseFailure(APIException):
+    status_code = 400
+    default_detail = "Error from third-party poll provider"
+
+    def __init__(self, detail):
+        self.detail = detail
+
+def Closed(): # noqa
+    end_date_params = {
+        "lte": timezone.now(),
+    }
+
+    return Range(end_date=end_date_params)
+
+class PollManager(ContentManager):
+    def search(self, **kwargs):
+        search_query = super(PollManager, self).search(**kwargs)
+
+        if "active" in kwargs:
+            search_query = search_query.filter(Published(before=timezone.now()))
+
+        if "closed" in kwargs:
+            search_query = search_query.filter(Closed())
+
+        return search_query
+
+class Poll(Content):
+    search_objects = PollManager()
+
+    question_text = models.TextField(blank=True, default="")
+    sodahead_id = models.CharField(max_length=20, blank=True, default="")
+    last_answer_index = models.IntegerField(default=0)
+    end_date = models.DateTimeField(null=True, default=None)
+    poll_image = ImageField(null=True, blank=True)
+
+    # This keeps Poll out of Content.search_objects
+    class Mapping(Content.Mapping):
+        poll_image = ElasticsearchImageField()
+        class Meta():
+            orphaned = True
+
+    def get_sodahead_data(self):
+        response = requests.get(SODAHEAD_POLL_ENDPOINT.format(self.sodahead_id))
+
+        if response.status_code != 200:
+            raise SodaheadResponseError(response.text)
+        else:
+            return response.json()
+
+    def sodahead_payload(self):
+        payload = {
+            'access_token': vault.read('sodahead/token')['value'],
+            'name': self.title,
+            'title': self.question_text,
+        }
+
+        if self.sodahead_id:
+            payload['id'] = self.sodahead_id
+
+        if self.published:
+            activation_date = self.published.astimezone(pytz.utc)
+            payload['activationDate'] = activation_date.strftime(SODAHEAD_DATE_FORMAT)
+
+        if self.end_date:
+            end_date = self.end_date.astimezone(pytz.utc)
+            payload['endDate'] = end_date.strftime(SODAHEAD_DATE_FORMAT)
+
+        for answer in self.answers.all():
+            if answer.answer_text == u'':
+                payload[answer.sodahead_answer_id] = BLANK_ANSWER
+            else:
+                payload[answer.sodahead_answer_id] = answer.answer_text
+
+        if not 'answer_01' in payload:
+            payload['answer_01'] = DEFAULT_ANSWER_1
+
+        if not 'answer_02' in payload:
+            payload['answer_02'] = DEFAULT_ANSWER_2
+
+        return payload
+
+
+    def save(self, *args, **kwargs):
+        if not self.sodahead_id:
+            response = requests.post(SODAHEAD_POLLS_ENDPOINT, self.sodahead_payload())
+            if response.status_code > 499:
```

Comment:
Looking at python-requests it seems like response.ok doesn't differentiate between client and server errors (4xx/5xx). Any suggestions based on that?